sqlalchemy_bigquery/_struct.py

```diff
@@ -0,0 +1,140 @@
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+from typing import Mapping, Tuple
+
+import packaging.version
+import sqlalchemy.sql.default_comparator
+import sqlalchemy.sql.sqltypes
+import sqlalchemy.types
+
+from . import base
+
+sqlalchemy_1_4_or_more = packaging.version.parse(
+    sqlalchemy.__version__
+) >= packaging.version.parse("1.4")
+
+if sqlalchemy_1_4_or_more:
+    import sqlalchemy.sql.coercions
+    import sqlalchemy.sql.roles
+
+# We have to delay getting the type compiler, because of circular imports. :(
+type_compiler = None
+
+
+class STRUCT(sqlalchemy.sql.sqltypes.Indexable, sqlalchemy.types.UserDefinedType):
+    """
+    A type for BigQuery STRUCT/RECORD data
+
+    See https://googleapis.dev/python/sqlalchemy-bigquery/latest/struct.html
+    """
+
+    # See https://docs.sqlalchemy.org/en/14/core/custom_types.html#creating-new-types
+
+    def __init__(
+        self,
+        *fields: Tuple[str, sqlalchemy.types.TypeEngine],
+        **kwfields: Mapping[str, sqlalchemy.types.TypeEngine],
+    ):
+        # Note that because:
+        # https://docs.python.org/3/whatsnew/3.6.html#pep-468-preserving-keyword-argument-order
+        # We know that `kwfields` preserves order.
+        self.__fields = tuple(
+            (
+                name,
+                type_ if isinstance(type_, sqlalchemy.types.TypeEngine) else type_(),
+            )
+            for (name, type_) in (fields + tuple(kwfields.items()))
+        )
+
+        self.__byname = {name.lower(): type_ for (name, type_) in self.__fields}
+
+    def __repr__(self):
+        fields = ", ".join(f"{name}={repr(type_)}" for name, type_ in self.__fields)
+        return f"STRUCT({fields})"
+
+    def get_col_spec(self, **kw):
+        global type_compiler
+
+        try:
+            process = type_compiler.process
+        except AttributeError:
+            type_compiler = base.dialect.type_compiler(base.dialect())
+            process = type_compiler.process
+
+        fields = ", ".join(f"{name} {process(type_)}" for name, type_ in self.__fields)
+        return f"STRUCT<{fields}>"
+
+    def bind_processor(self, dialect):
+        return dict
+
+    class Comparator(sqlalchemy.sql.sqltypes.Indexable.Comparator):
+        def _setup_getitem(self, name):
+            if not isinstance(name, str):
+                raise TypeError(
+                    f"STRUCT fields can only be accessed with strings field names,"
+                    f" not {name}."
+                )
+            subtype = self.expr.type._STRUCT__byname.get(name.lower())
```

Comment:
I refactored this to make name mangling more explicit and consistent, so I don't think comments are needed anymore. See if you agree. :)

I mainly use "private" variables, which aren't :), to avoid namespace conflicts when subclassing across responsibility boundaries. Arguably, explicit naming is better.